nbodykit/io/csv.py

```diff
@@ -1,17 +1,136 @@
 import numpy
-
+import os
+from pandas import read_csv 
+    
 from ..extern.six import string_types
 from . import FileType, tools
 
+class CSVPartition(object):
+    """
+    A simple class to convert byte strings of data from a CSV file
+    to a pandas DataFrame on demand
+    
+    The DataFrame is cached as :attr:`value`, so only a single
+    call to :func:`pandas.read_csv` is used
+    """
+    def __init__(self, bstr, **config):
+        """
+        Parameters
+        ----------
+        bstr : bytestring
+            the data content read from file, as a string of bytes
+        config : 
+            the configuration keywords passed to :func:`pandas.read_csv`
+        """
+        self.bstr = bstr
+        self.config = config
+        
+    @property
+    def value(self):
+        """
+        Return the parsed btye string as a DataFrame
+        """
+        try:
+            return self._value
+        except AttributeError:
+            from io import BytesIO
+            
+            # parse the byte string
+            b = BytesIO()
+            b.write(self.bstr); b.seek(0)
+            self._value = read_csv(b, **self.config)
+            
+            # free memory, since we have DataFrame now
+            self.bstr = None 
+            
+            return self._value
+
+def make_partitions(filename, blocksize, config, delimiter="\n"):
+    """
+    Partition a CSV file into blocks, using the preferred blocksize 
+    in bytes, returning the partititions and number of rows in 
+    each partition
+
+    This divides the input file into partitions with size
+    roughly equal to blocksize, reads the bytes, and counts
+    the number of delimiters to compute the size of each block
+    
+    Parameters
+    ----------
+    filename : str
+        the name of the CSV file to load
+    blocksize : int
+        the desired number of bytes per block 
+    delimiter : str, optional
+        the character separating lines; default is
+        the newline character
+    config : dict
+        any keyword options to pass to :func:`pandas.read_csv`
+    
+    Returns
+    -------
+    partitions : list of CSVPartition
+        list of objects storing the data content of each file partition,
+        stored as a bytestring
+    sizes : list of int
+        the list of the number of rows in each partition
+    """
+    from dask.bytes.utils import read_block
```

Comment:
@rainwoodman this is the only dependency from dask still...it handles partitioning the file into blocks, always ending on the newline character. This function seems pretty stable to me...